### PR TITLE
Figure out if build has compatible standard set

### PR DESCRIPTION
**Pull request by paulgessinger**
Currently `DD4hepConfig.cmake`, when included via `find_package`, will silently modify the build environment of the including cmake project. I reckon this is to try to harmonize with ROOT's locking in of the C++ standards version in their baked RStringView.cxx.

However, silently modifying the including projects settings seems intrusive.

First of all: is that really what is happening, or am I missing something?

If that is actually the behaviour, this PR checks if the standards version is explicitly set, and if so, compares against what DD4hep was built with (using the `@CMAKE_CXX_STANDARD@` thats replaced in the config template). If those do not match, print a visible error warning, and **don't modify CMAKE_CXX_STANDARD**. This will cause a compiler error eventually.

BEGINRELEASENOTES

* In DD4hepConfig.cmake, figure out if build has compatible standard set and print error if not
 
ENDRELEASENOTES

### CODE REVIEW

File: cmake/DD4hepConfig.cmake.in [hunk 1]
```diff
@@ -21,12 +21,24 @@ if (_lastdir STREQUAL "cmake")
...
+if("${CMAKE_CXX_STANDARD}" STREQUAL "")
+  set ( CMAKE_CXX_STANDARD  "${DD4hep_BUILD_CXX_STANDARD}" )
+else()
+  if (NOT "${CMAKE_CXX_STANDARD}" EQUAL "${DD4hep_BUILD_CXX_STANDARD}")
+    message( SEND_ERROR "CMAKE_CXX_STANDARD is set to C++${CMAKE_CXX_STANDARD}\
+             but C++${DD4hep_BUILD_CXX_STANDARD} is required.")
+  endif()
+endif()
+
 set ( CMAKE_CXX_STANDARD  @CMAKE_CXX_STANDARD@ )
```
Comment (andresailer) [inline cmake/DD4hepConfig.cmake.in]: You would want to remove this line as well, right?

File: cmake/DD4hepConfig.cmake.in [hunk 2]
```diff
@@ -21,12 +21,24 @@ if (_lastdir STREQUAL "cmake")
...
+set ( DD4HEP_USE_XERCESC        "@DD4HEP_USE_XERCESC@" )
+set ( DD4hep_BUILD_CXX_STANDARD @CMAKE_CXX_STANDARD@ )
+
+# figure out if including build has compatible standard set
+if("${CMAKE_CXX_STANDARD}" STREQUAL "")
+  set ( CMAKE_CXX_STANDARD  "${DD4hep_BUILD_CXX_STANDARD}" )
+else()
+  if (NOT "${CMAKE_CXX_STANDARD}" EQUAL "${DD4hep_BUILD_CXX_STANDARD}")
+    message( SEND_ERROR "CMAKE_CXX_STANDARD is set to C++${CMAKE_CXX_STANDARD}\
+             but C++${DD4hep_BUILD_CXX_STANDARD} is required.")
```
Comment (andresailer) [inline cmake/DD4hepConfig.cmake.in]: Maybe add a DD4hep in the error message to make clear were this comes from?

Comment (paulgessinger) [reply cmake/DD4hepConfig.cmake.in]: Right, will change that.

File: cmake/DD4hepConfig.cmake.in (hunk 1 above)
Comment (paulgessinger) [reply cmake/DD4hepConfig.cmake.in]: Absolutely, I guess I missed that.

**Pull request merged**